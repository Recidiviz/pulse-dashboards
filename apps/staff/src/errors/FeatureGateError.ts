// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

/**
 * Exception to be thrown when a feature flag forbids progress.
 * Useful in situations where the flag cannot be checked until after data is fetched
 * and you need to abort hydration when the flag is turned off.
 */
export class FeatureGateError extends Error {}
