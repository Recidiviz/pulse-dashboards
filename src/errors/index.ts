export { FeatureGateError } from "./FeatureGateError";
export { OpportunityValidationError } from "./OpportunityValidationError";
export { TaskValidationError } from "./TaskValidationError";
