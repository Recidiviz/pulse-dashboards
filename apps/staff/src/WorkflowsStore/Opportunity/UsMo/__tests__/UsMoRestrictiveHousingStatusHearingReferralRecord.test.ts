// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2024 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================

import tk from "timekeeper";

import { OpportunityValidationError } from "../../../../errors";
import {
  UsMoRestrictiveHousingStatusHearingReferralRecordRaw,
  usMoRestrictiveHousingStatusHearingSchema,
  validateReferral,
} from "../UsMoRestrictiveHousingStatusHearingOpportunity";

const rawRecordBase: Omit<
  UsMoRestrictiveHousingStatusHearingReferralRecordRaw,
  "eligibleCriteria" | "ineligibleCriteria"
> = {
  stateCode: "US_MO",
  externalId: "004",
  metadata: {
    mostRecentHearingDate: "2022-09-03",
    mostRecentHearingType: "hearing type",
    mostRecentHearingFacility: "FACILITY NAME",
    mostRecentHearingComments: "Reason for Hearing: 30 day review",
    currentFacility: "FACILITY 01",
    restrictiveHousingStartDate: "2022-10-01",
    bedNumber: "03",
    roomNumber: "05",
    complexNumber: "2",
    buildingNumber: "13",
    housingUseCode: "HOS",
    majorCdvs: [
      {
        cdvDate: "2022-02-20",
        cdvRule: "Rule 7.2",
      },
    ],
    cdvsSinceLastHearing: [],
    numMinorCdvsBeforeLastHearing: "3",
  },
};

const rawRecordUpcomingHearingToday: UsMoRestrictiveHousingStatusHearingReferralRecordRaw =
  {
    ...rawRecordBase,
    eligibleCriteria: {
      usMoInRestrictiveHousing: {
        confinementType: "confinement type",
      },
    },
    ineligibleCriteria: {
      usMoOverdueForHearing: {
        nextReviewDate: "2023-11-03", // This should match "today" defined below
      },
    },
  };

const rawRecordOverdue: UsMoRestrictiveHousingStatusHearingReferralRecordRaw = {
  ...rawRecordBase,
  eligibleCriteria: {
    usMoOverdueForHearing: {
      nextReviewDate: "2023-10-03",
    },
    usMoInRestrictiveHousing: {
      confinementType: "confinement type",
    },
  },
  ineligibleCriteria: {},
};

const rawRecordMissingHearingDate: UsMoRestrictiveHousingStatusHearingReferralRecordRaw =
  {
    ...rawRecordBase,
    eligibleCriteria: {
      usMoInRestrictiveHousing: {
        confinementType: "confinement type",
      },
    },
    ineligibleCriteria: {
      usMoOverdueForHearing: {
        nextReviewDate: null,
      },
    },
  };

const rawRecordMissingHearingDateCriteria: UsMoRestrictiveHousingStatusHearingReferralRecordRaw =
  {
    ...rawRecordBase,
    eligibleCriteria: {
      usMoInRestrictiveHousing: {
        confinementType: "confinement type",
      },
    },
    ineligibleCriteria: {
      usMoOverdueForHearing: null,
    },
  };

const rawRecordHearingDateFarInFuture: UsMoRestrictiveHousingStatusHearingReferralRecordRaw =
  {
    ...rawRecordBase,
    eligibleCriteria: {
      usMoInRestrictiveHousing: {
        confinementType: "confinement type",
      },
    },
    ineligibleCriteria: {
      usMoOverdueForHearing: {
        nextReviewDate: "2024-01-01",
      },
    },
  };

const rawRecordIneligibleButDateInPast: UsMoRestrictiveHousingStatusHearingReferralRecordRaw =
  {
    ...rawRecordBase,
    eligibleCriteria: {
      usMoInRestrictiveHousing: {
        confinementType: "confinement type",
      },
    },
    ineligibleCriteria: {
      usMoOverdueForHearing: {
        nextReviewDate: "2023-10-02",
      },
    },
  };

const today = new Date(2023, 10, 3);

beforeEach(() => {
  tk.freeze(today);
});

afterEach(() => {
  tk.reset();
});

test("transform record", () => {
  expect(
    usMoRestrictiveHousingStatusHearingSchema.parse(
      rawRecordUpcomingHearingToday,
    ),
  ).toMatchSnapshot();
  expect(
    usMoRestrictiveHousingStatusHearingSchema.parse(rawRecordOverdue),
  ).toMatchSnapshot();
  expect(
    usMoRestrictiveHousingStatusHearingSchema.parse(
      rawRecordMissingHearingDate,
    ),
  ).toMatchSnapshot();
  expect(
    usMoRestrictiveHousingStatusHearingSchema.parse(
      rawRecordMissingHearingDateCriteria,
    ),
  ).toMatchSnapshot();
});

test("transform ineligible record with hearing date in past into eligible criteria", () => {
  const parsedRecord = usMoRestrictiveHousingStatusHearingSchema.parse(
    rawRecordIneligibleButDateInPast,
  );
  expect(parsedRecord).toMatchSnapshot();
  expect(parsedRecord.ineligibleCriteria).toBeEmpty();
  expect(
    parsedRecord.eligibleCriteria.usMoOverdueForHearing?.nextReviewDate,
  ).toBeDefined();
});

test("record validates: overdue for hearing in the past", () => {
  expect(() =>
    validateReferral(
      usMoRestrictiveHousingStatusHearingSchema.parse(rawRecordOverdue),
    ),
  ).not.toThrow(OpportunityValidationError);
});

test("record validates: next review date today", () => {
  expect(() =>
    validateReferral(
      usMoRestrictiveHousingStatusHearingSchema.parse(
        rawRecordUpcomingHearingToday,
      ),
    ),
  ).not.toThrow(OpportunityValidationError);
});

test("record validates: missing hearing date", () => {
  expect(() =>
    validateReferral(
      usMoRestrictiveHousingStatusHearingSchema.parse(
        rawRecordMissingHearingDate,
      ),
    ),
  ).not.toThrow(OpportunityValidationError);
});

test("record validates: hearing date far in future", () => {
  expect(() =>
    validateReferral(
      usMoRestrictiveHousingStatusHearingSchema.parse(
        rawRecordHearingDateFarInFuture,
      ),
    ),
  ).not.toThrow(OpportunityValidationError);
});

test("record does not validate: overdue in the future", () => {
  const overdueInFuture: UsMoRestrictiveHousingStatusHearingReferralRecordRaw =
    {
      ...rawRecordOverdue,
      eligibleCriteria: {
        usMoOverdueForHearing: {
          nextReviewDate: "3000-01-01",
        },
        usMoInRestrictiveHousing: {
          confinementType: "confinement type",
        },
      },
    };
  expect(() =>
    validateReferral(
      usMoRestrictiveHousingStatusHearingSchema.parse(overdueInFuture),
    ),
  ).toThrow(OpportunityValidationError);
});
