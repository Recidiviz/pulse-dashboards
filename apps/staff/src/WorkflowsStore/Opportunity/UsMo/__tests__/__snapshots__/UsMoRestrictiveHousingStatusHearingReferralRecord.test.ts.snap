// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transform ineligible record with hearing date in past into eligible criteria 1`] = `
{
  "caseNotes": {},
  "eligibleCriteria": {
    "usMoInRestrictiveHousing": {
      "confinementType": "confinement type",
    },
    "usMoOverdueForHearing": {
      "nextReviewDate": 2023-10-02T00:00:00.000Z,
    },
  },
  "externalId": "004",
  "ineligibleCriteria": {},
  "metadata": {
    "bedNumber": "03",
    "buildingNumber": "13",
    "cdvsSinceLastHearing": [],
    "complexNumber": "2",
    "currentFacility": "FACILITY 01",
    "housingUseCode": "HOS",
    "majorCdvs": [
      {
        "cdvDate": 2022-02-20T00:00:00.000Z,
        "cdvRule": "Rule 7.2",
      },
    ],
    "mostRecentHearingComments": "Reason for Hearing: 30 day review",
    "mostRecentHearingDate": 2022-09-03T00:00:00.000Z,
    "mostRecentHearingFacility": "FACILITY NAME",
    "mostRecentHearingType": "hearing type",
    "numMinorCdvsBeforeLastHearing": 3,
    "restrictiveHousingStartDate": 2022-10-01T00:00:00.000Z,
    "roomNumber": "05",
  },
  "stateCode": "US_MO",
}
`;

exports[`transform record 1`] = `
{
  "caseNotes": {},
  "eligibleCriteria": {
    "usMoInRestrictiveHousing": {
      "confinementType": "confinement type",
    },
  },
  "externalId": "004",
  "ineligibleCriteria": {
    "usMoOverdueForHearing": {
      "nextReviewDate": 2023-11-03T00:00:00.000Z,
    },
  },
  "metadata": {
    "bedNumber": "03",
    "buildingNumber": "13",
    "cdvsSinceLastHearing": [],
    "complexNumber": "2",
    "currentFacility": "FACILITY 01",
    "housingUseCode": "HOS",
    "majorCdvs": [
      {
        "cdvDate": 2022-02-20T00:00:00.000Z,
        "cdvRule": "Rule 7.2",
      },
    ],
    "mostRecentHearingComments": "Reason for Hearing: 30 day review",
    "mostRecentHearingDate": 2022-09-03T00:00:00.000Z,
    "mostRecentHearingFacility": "FACILITY NAME",
    "mostRecentHearingType": "hearing type",
    "numMinorCdvsBeforeLastHearing": 3,
    "restrictiveHousingStartDate": 2022-10-01T00:00:00.000Z,
    "roomNumber": "05",
  },
  "stateCode": "US_MO",
}
`;

exports[`transform record 2`] = `
{
  "caseNotes": {},
  "eligibleCriteria": {
    "usMoInRestrictiveHousing": {
      "confinementType": "confinement type",
    },
    "usMoOverdueForHearing": {
      "nextReviewDate": 2023-10-03T00:00:00.000Z,
    },
  },
  "externalId": "004",
  "ineligibleCriteria": {},
  "metadata": {
    "bedNumber": "03",
    "buildingNumber": "13",
    "cdvsSinceLastHearing": [],
    "complexNumber": "2",
    "currentFacility": "FACILITY 01",
    "housingUseCode": "HOS",
    "majorCdvs": [
      {
        "cdvDate": 2022-02-20T00:00:00.000Z,
        "cdvRule": "Rule 7.2",
      },
    ],
    "mostRecentHearingComments": "Reason for Hearing: 30 day review",
    "mostRecentHearingDate": 2022-09-03T00:00:00.000Z,
    "mostRecentHearingFacility": "FACILITY NAME",
    "mostRecentHearingType": "hearing type",
    "numMinorCdvsBeforeLastHearing": 3,
    "restrictiveHousingStartDate": 2022-10-01T00:00:00.000Z,
    "roomNumber": "05",
  },
  "stateCode": "US_MO",
}
`;

exports[`transform record 3`] = `
{
  "caseNotes": {},
  "eligibleCriteria": {
    "usMoInRestrictiveHousing": {
      "confinementType": "confinement type",
    },
  },
  "externalId": "004",
  "ineligibleCriteria": {
    "usMoOverdueForHearing": {
      "nextReviewDate": undefined,
    },
  },
  "metadata": {
    "bedNumber": "03",
    "buildingNumber": "13",
    "cdvsSinceLastHearing": [],
    "complexNumber": "2",
    "currentFacility": "FACILITY 01",
    "housingUseCode": "HOS",
    "majorCdvs": [
      {
        "cdvDate": 2022-02-20T00:00:00.000Z,
        "cdvRule": "Rule 7.2",
      },
    ],
    "mostRecentHearingComments": "Reason for Hearing: 30 day review",
    "mostRecentHearingDate": 2022-09-03T00:00:00.000Z,
    "mostRecentHearingFacility": "FACILITY NAME",
    "mostRecentHearingType": "hearing type",
    "numMinorCdvsBeforeLastHearing": 3,
    "restrictiveHousingStartDate": 2022-10-01T00:00:00.000Z,
    "roomNumber": "05",
  },
  "stateCode": "US_MO",
}
`;

exports[`transform record 4`] = `
{
  "caseNotes": {},
  "eligibleCriteria": {
    "usMoInRestrictiveHousing": {
      "confinementType": "confinement type",
    },
  },
  "externalId": "004",
  "ineligibleCriteria": {
    "usMoOverdueForHearing": {},
  },
  "metadata": {
    "bedNumber": "03",
    "buildingNumber": "13",
    "cdvsSinceLastHearing": [],
    "complexNumber": "2",
    "currentFacility": "FACILITY 01",
    "housingUseCode": "HOS",
    "majorCdvs": [
      {
        "cdvDate": 2022-02-20T00:00:00.000Z,
        "cdvRule": "Rule 7.2",
      },
    ],
    "mostRecentHearingComments": "Reason for Hearing: 30 day review",
    "mostRecentHearingDate": 2022-09-03T00:00:00.000Z,
    "mostRecentHearingFacility": "FACILITY NAME",
    "mostRecentHearingType": "hearing type",
    "numMinorCdvsBeforeLastHearing": 3,
    "restrictiveHousingStartDate": 2022-10-01T00:00:00.000Z,
    "roomNumber": "05",
  },
  "stateCode": "US_MO",
}
`;
